plugin/src/main/java/org/opensearch/ml/model/MLModelAutoReloader.java

```diff
@@ -0,0 +1,329 @@
+import org.opensearch.ml.utils.MLNodeUtils;
+import org.opensearch.rest.RestStatus;
+import org.opensearch.search.SearchHit;
+import org.opensearch.search.builder.SearchSourceBuilder;
+import org.opensearch.search.sort.FieldSortBuilder;
+import org.opensearch.search.sort.SortBuilder;
+import org.opensearch.search.sort.SortOrder;
+import org.opensearch.threadpool.ThreadPool;
+
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * Manager class for ML models and nodes. It contains ML model auto reload operations etc.
+ */
+@Log4j2
+public class MLModelAutoReloader {
+
+    private final Client client;
+    private final ClusterService clusterService;
+    private final NamedXContentRegistry xContentRegistry;
+    private final DiscoveryNodeHelper nodeHelper;
+    private final ThreadPool threadPool;
+    private volatile Boolean enableAutoReloadModel;
+    private volatile Integer autoReloadMaxRetryTimes;
+
+    /**
+     * constructor method， init all the params necessary for model auto reloading
+     *
+     * @param clusterService   clusterService
+     * @param threadPool       threadPool
+     * @param client           client
+     * @param xContentRegistry xContentRegistry
+     * @param nodeHelper       nodeHelper
+     * @param settings         settings
+     */
+    public MLModelAutoReloader(
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        Client client,
+        NamedXContentRegistry xContentRegistry,
+        DiscoveryNodeHelper nodeHelper,
+        Settings settings
+    ) {
+        this.clusterService = clusterService;
+        this.client = client;
+        this.xContentRegistry = xContentRegistry;
+        this.nodeHelper = nodeHelper;
+        this.threadPool = threadPool;
+
+        enableAutoReloadModel = ML_COMMONS_MODEL_AUTO_RELOAD_ENABLE.get(settings);
+        autoReloadMaxRetryTimes = ML_MODEL_RELOAD_MAX_RETRY_TIMES.get(settings);
+        clusterService
+            .getClusterSettings()
+            .addSettingsUpdateConsumer(ML_COMMONS_MODEL_AUTO_RELOAD_ENABLE, it -> enableAutoReloadModel = it);
+
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(ML_MODEL_RELOAD_MAX_RETRY_TIMES, it -> autoReloadMaxRetryTimes = it);
+    }
+
+    /**
+     * the main method: model auto reloading
+     */
+    public void autoReloadModel() {
+        log.info("auto reload model enabled: {} ", enableAutoReloadModel);
+
+        // if we don't need to reload automatically, just return without doing anything
+        if (!enableAutoReloadModel) {
+            return;
+        }
+
+        // At opensearch startup, get local node id, if not ml node,we ignored, just return without doing anything
+        if (!MLNodeUtils.isMLNode(clusterService.localNode())) {
+            return;
+        }
+
+        String localNodeId = clusterService.localNode().getId();
+        // auto reload all models of this local ml node
+        threadPool.generic().submit(() -> {
```

Comment:
Please change this threadpool to `UPLOAD_THREAD_POOL` in MachineLearningPlugin since this is dedicated for uploading models.

Reply:
en ,yaliang have said it~ I have committed the latest code